libraries-3/src/main/java/com/baeldung/jnats/NatsClient.java

```diff
@@ -1,123 +1,142 @@
 package com.baeldung.jnats;
 
+import io.nats.client.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.io.IOException;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+public final class NatsClient implements AutoCloseable {
 
-import io.nats.client.AsyncSubscription;
-import io.nats.client.Connection;
-import io.nats.client.Message;
-import io.nats.client.Nats;
-import io.nats.client.Options;
-import io.nats.client.Subscription;
-import io.nats.client.SyncSubscription;
+    private final Map<Subscription, Dispatcher> dispatcherBySubscription = new HashMap<>();
 
-public final class NatsClient {
+    private final static Logger log = LoggerFactory.getLogger(NatsClient.class);
 
     private final String serverURI;
 
     private final Connection natsConnection;
 
-    private final Map<String, Subscription> subscriptions = new HashMap<>();
-
-    private final static Logger log = LoggerFactory.getLogger(NatsClient.class);
-
-    NatsClient() {
-        this.serverURI = "jnats://localhost:4222";
-        natsConnection = initConnection(serverURI);
+    public NatsClient() throws IOException, InterruptedException {
+        this(null);
     }
 
-    public NatsClient(String serverURI) {
-        if ((serverURI != null) && (!serverURI.isEmpty())) {
+    public NatsClient(String serverURI) throws IOException, InterruptedException {
+        if (serverURI == null || serverURI.isEmpty()) {
+            this.serverURI = "nats://localhost:4222";
+        }
+        else {
```

Comment:
This is what I mean by a static factory method and a single responsibility constructor:
```suggestion
    private NatsClient(String serverURI, Connection natsConnection) {
        this.serverURI = serverURI;
        this. natsConnection = natsConnection;
    }
    
    public static NatsClient connectNatsClient(String serverURI) throws IOException, InterruptedException {
        Connection natsConnection = createConnection(serverURI);
        return new NatsClient(serverURI, natsConnection);
    }
```

I also think it's clearer if we remove the empty serverURI logic and make that explicit in the test

Reply:
This is logic in the constructor, so I'm not sure what you mean.
```
Connection natsConnection = createConnection(serverURI);
```
Anyway, I changed it to only take a Connection in the constructor and provided a static method to make a connection. This is probably a more normal way this client would be done anyway, since making a connection can get very involved.